### DISCUSSION TITLE
OpenGL on iOS: Black screen

### DISCUSSION THREAD

**Discussion by davidgiga1993** [AvaloniaUI/Avalonia#17014] [Q&A]
### Describe the bug

OpenGlControlBase doesn't render anything on iOS, the debug overlays frame counter also do not count upwards, as if no frame is being rendered.

### To Reproduce

Run ControlCatalog sample app on iOS

### Expected behavior

openGL should be rendered (ideally via ANGLE).

### Avalonia version

11.2.0-beta1

### OS

iOS

**Comment by kekekeks** [comment]
Are you checking with actual device or simulator?

**Comment by kekekeks** [comment]
Also, you need to use OpenGL rendering mode for the entire app (Metal is used by default)

**Comment by davidgiga1993** [comment]
Tested both, any example how to set the rendering mode? I don't find anything related in the documentation. I suspect it's something with `UseRenderingSubsystem`?

Update: Found it (for future reference):
```csharp
protected override AppBuilder CustomizeAppBuilder(AppBuilder builder)
  {
      return base.CustomizeAppBuilder(builder)
          .With(new iOSPlatformOptions{RenderingMode = [iOSRenderingMode.OpenGl]})
          .WithInterFont();
  }
```

Update 2:
`WithInterFont()` needs to be removed, otherwise the opengl change crashes the app.

I see the renderer is listed as `Renderer: Apple M1 GPU Version: OpenGL ES 3.0 Metal - 101` does that mean it's using the deprecated openGL api and not ANGLE?

**Comment by maxkatz6** [comment]
WithInterFont shouldn't really have any effect on rendering pipeline.

> does that mean it's using the deprecated openGL api and not ANGLE?

Yes. Which is why Metal is enabled by default now, and not OpenGL.

**Reply by davidgiga1993** [reply]
I know, but it crashes in opengl mode when interfont is enabled, not sure why.

Is there a way to use angle on iOS? I'm asking as we all don't know how long Apple will continue including OpenGL drivers and this drives the decision which framework to use for best long-term platform support (with opengl)

**Reply by orosbogdan** [reply]
> I know, but it crashes in opengl mode when interfont is enabled, not sure why.
> 
> Is there a way to use angle on iOS? I'm asking as we all don't know how long Apple will continue including OpenGL drivers and this drives the decision which framework to use for best long-term platform support (with opengl)

Did you find a way to do it?

**Comment by orosbogdan** [comment]
Is this fixed ?

**Comment by maxhoerstr** [comment]
Hey, is there anything new to this issue? I tried to use the openGL rendering mode but my app crashes with and without the "WithInterFont" decorator.

**Reply by maxkatz6** [reply]
No. Metal is default, because it's stable and supported by Apple.
But Avalonia doesn't have integration between Metal and OpenGL yet.
Native OpenGL is disabled by default, because it's broken on latest simulators, with reasons outside of the avalonia, it still should work on physical devices though.

**Reply by maxkatz6** [reply]
> Avalonia doesn't have integration between Metal and OpenGL yet

This would be a proper solution on our side.
Using Metal for the app and only use gl via IOSurface when requested by developer. 
But it's not prioritized.

**Reply by maxhoerstr** [reply]
The crash was due to some weird dotnet stuff which i fixed. Now, i can load the app but see a blank screen. The renderer printed is metal. I load a 3D Model in my app via silk assimp and then use opengl to render it. This works on macOS with metal, however, i see nothing on my iPad. Should it work on my device?

**Reply by maxkatz6** [reply]
It won't work with Metal in current implementation. Only if you switch to OpenGL rendering mode.

**Reply by maxhoerstr** [reply]
Okay but even if i set the rendering mode to OpenGL it still prints Metal as renderer.

**Reply by anose001** [reply]
@maxhoerstr , have you been able to get this running on iOS?
